src/main/java/net/neoforged/neoforge/common/crafting/outgredient/TagDefaultsManager.java

```diff
@@ -0,0 +1,106 @@
+import net.minecraft.server.packs.resources.SimpleJsonResourceReloadListener;
+import net.minecraft.tags.TagKey;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.util.profiling.ProfilerFiller;
+import net.neoforged.neoforge.common.NeoForgeEventHandler;
+import net.neoforged.neoforge.common.config.NeoForgeServerConfig;
+import org.slf4j.Logger;
+
+/**
+ * This is the central manager class for {@link TagDefaults}. Access an instance via {@link NeoForgeEventHandler#getTagDefaultsManager()}.
+ */
+public class TagDefaultsManager extends SimpleJsonResourceReloadListener<JsonElement> {
+    public static final String FOLDER = "tag_defaults";
+    private static final Logger LOGGER = LogUtils.getLogger();
+    private final Map<ResourceKey<? extends Registry<?>>, TagDefaults<?>> entries = new HashMap<>();
+
+    public TagDefaultsManager() {
+        super(ExtraCodecs.JSON, FileToIdConverter.json(FOLDER));
+    }
+
+    @Override
+    protected void apply(Map<ResourceLocation, JsonElement> resourceList, ResourceManager resourceManager, ProfilerFiller profilerFiller) {
+        entries.clear();
+        resourceList.forEach(this::put);
+    }
+
+    /**
+     * Resolves a {@link TagDefaults} into an optional outgredient.
+     *
+     * @param registryKey The key of the associated {@link Registry}.
+     * @param tagKey      The {@link TagKey} to resolve.
+     * @return An {@link Optional} of a {@link TagDefaults}'s resolved value, or an empty {@link Optional} if the {@link TagDefaults} could not be resolved.
+     * @param <T> The type of the {@link Registry} and the outgredient.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> Optional<T> resolve(ResourceKey<? extends Registry<T>> registryKey, TagKey<T> tagKey) {
+        // Try to find an entry in the tag defaults themselves.
+        TagDefaults<T> defaults = (TagDefaults<T>) entries.get(registryKey);
+        if (defaults != null) {
+            T result = defaults.resolve(tagKey);
+            if (result != null) return Optional.of(result);
+        }
+
+        // Try to find an entry by walking the config mod id list.
+        Registry<T> registry = (Registry<T>) BuiltInRegistries.REGISTRY.getValue(registryKey.location());
+        if (registry != null) {
+            List<String> modIds = NeoForgeServerConfig.INSTANCE.defaultedTagModIds.get();
+            // Walk the config mod id list.
+            for (String modId : modIds) {
+                // Filter the tag contents by mod id.
+                List<Holder<T>> tagContents = registry
+                        .getOrThrow(tagKey)
+                        .stream()
+                        .filter(holder -> holder.unwrapKey().map(key -> key.location().getNamespace().equals(modId)).orElse(false))
+                        .toList();
+                // If we have exactly one candidate, we found our result.
+                if (tagContents.size() == 1) return Optional.of(tagContents.getFirst().value());
```

Comment:
If we have multiple, just use the first. Exactly one is going to be extremely unstable, most notably when it comes to fluids. If you want, reference my `TagPreference` for how to do this in a stable way: https://github.com/SlimeKnights/Mantle/blob/1.20/src/main/java/slimeknights/mantle/recipe/helper/TagPreference.java

Also, cache this. No reason we should be iterating the tag every time someone asks for a tag default. Recipe results get queried many times and have no reason to change until datapack reload.